File: src/conduit/engine/extract.py
"""Extract data from sources into Arrow tables."""

from __future__ import annotations

import logging
from pathlib import Path

import pyarrow as pa
import pyarrow.csv as pcsv

from conduit.models import SourceConfig

logger = logging.getLogger(__name__)

# Extractors keyed by source type
_EXTRACTORS: dict[str, callable] = {}


def _extract_csv(source: SourceConfig, base_dir: Path) -> pa.Table:
    file_path = base_dir / source.config.get("path", "")
    if not file_path.exists():
        raise FileNotFoundError(f"CSV source file not found: {file_path}")

    read_options = pcsv.ReadOptions()
    if not source.config.get("has_header", True):
        read_options.autogenerate_column_names = True

    parse_options = pcsv.ParseOptions()
    delimiter = source.config.get("delimiter", ",")
    if delimiter:
        parse_options.delimiter = delimiter

    table = pcsv.read_csv(str(file_path), read_options=read_options, parse_options=parse_options)
    logger.info("Extracted %d rows from CSV source '%s' (%s)", len(table), source.name, file_path)
    return table


def _extract_tsv(source: SourceConfig, base_dir: Path) -> pa.Table:
    source.config.setdefault("delimiter", "\t")
    return _extract_csv(source, base_dir)


_EXTRACTORS = {
    "csv": _extract_csv,
    "tsv": _extract_tsv,
}


def extract_sources(sources: list[SourceConfig], base_dir: Path) -> dict[str, pa.Table]:
    """Extract all sources and return a dict mapping source name to Arrow table."""
    tables: dict[str, pa.Table] = {}

    for source in sources:
        extractor = _EXTRACTORS.get(source.type)
        if extractor is None:
            raise NotImplementedError(
                f"Connector type '{source.type}' is not yet implemented. "
                f"Supported types: {', '.join(sorted(_EXTRACTORS))}"
            )
        tables[source.name] = extractor(source, base_dir)

    logger.info("Extracted %d source(s): %s", len(tables), ", ".join(tables))
    return tables

File: src/ts/engine/extract.ts
/**
 * Extract data from sources into DataTables.
 */

import { parse as parseCsv } from "@std/csv";
import { join } from "@std/path";
import type { DataTable, SourceConfig } from "../models.ts";

type ExtractFn = (
  source: SourceConfig,
  baseDir: string,
) => DataTable;

function extractCsv(source: SourceConfig, baseDir: string): DataTable {
  const filePath = join(baseDir, (source.config.path as string) ?? "");

  try {
    Deno.statSync(filePath);
  } catch {
    throw new Error(`CSV source file not found: ${filePath}`);
  }

  const text = Deno.readTextFileSync(filePath);
  const delimiter = (source.config.delimiter as string) ?? ",";
  const hasHeader = (source.config.has_header as boolean) ?? true;

  const parsed = parseCsv(text, {
    skipFirstRow: hasHeader,
    separator: delimiter,
  });

  const rows = parsed as Record<string, unknown>[];
  const columns = rows.length > 0
    ? Object.keys(rows[0]!).map((name) => ({ name, type: "VARCHAR" }))
    : [];

  console.log(
    `Extracted ${rows.length} rows from CSV source '${source.name}' (${filePath})`,
  );

  return { columns, rows };
}

function extractTsv(source: SourceConfig, baseDir: string): DataTable {
  if (!source.config.delimiter) {
    source.config.delimiter = "\t";
  }
  return extractCsv(source, baseDir);
}

export const EXTRACTORS: Map<string, ExtractFn> = new Map([
  ["csv", extractCsv],
  ["tsv", extractTsv],
]);

export function extractSources(
  sources: SourceConfig[],
  baseDir: string,
): Map<string, DataTable> {
  const tables = new Map<string, DataTable>();

  for (const source of sources) {
    const extractor = EXTRACTORS.get(source.type);
    if (!extractor) {
      throw new Error(
        `Connector type '${source.type}' is not yet implemented. ` +
          `Supported types: ${[...EXTRACTORS.keys()].sort().join(", ")}`,
      );
    }
    tables.set(source.name, extractor(source, baseDir));
  }

  console.log(
    `Extracted ${tables.size} source(s): ${[...tables.keys()].join(", ")}`,
  );
  return tables;
}
